operator/src/lib.rs

```diff
@@ -94,101 +163,103 @@ impl KafkaState {
-        // broker from the spec, we then delete it.
-        for pod in &self.existing_pods {
-            if !self
-                .kafka_spec
-                .brokers
-                .iter()
-                .any(|broker| podutils::is_pod_assigned_to_node(pod, &broker.node_name))
-            {
-                self.context.client.delete(pod).await?;
-                return Ok(ReconcileFunctionAction::Requeue(Duration::from_secs(10)));
+    async fn create_missing_pods(&mut self) -> KafkaReconcileResult {
+        // The iteration happens in two stages here, to accomodate the way our operators think
+        // about nodes and roles.
+        // The hierarchy is:
+        // - Roles (for example Datanode, Namenode, Kafka Broker)
+        //   - Node groups for this role (user defined)
+        //      - Individual nodes
+        for node_type in KafkaNodeType::iter() {
+            if let Some(nodes_for_role) = self.eligible_nodes.get(&node_type) {
+                for (role_group, nodes) in nodes_for_role {
+                    // Create config map for this rolegroup
+                    let pod_name =
+                        format!("kafka-{}-{}-{}", self.context.name(), role_group, node_type)
+                            .to_lowercase();
+                    let cm_name = format!("{}-config", pod_name);
+                    debug!("pod_name: [{}], cm_name: [{}]", pod_name, cm_name);
+
+                    self.create_config_map(&cm_name).await?;
+                    debug!(
+                        "Identify missing pods for [{}] role and group [{}]",
+                        node_type, role_group
+                    );
+                    trace!(
+                        "candidate_nodes[{}]: [{:?}]",
+                        nodes.len(),
+                        nodes
+                            .iter()
+                            .map(|node| node.metadata.name.clone().unwrap())
+                            .collect::<Vec<_>>()
+                    );
+                    trace!(
+                        "existing_pods[{}]: [{:?}]",
+                        &self.existing_pods.len(),
+                        &self
+                            .existing_pods
+                            .iter()
+                            .map(|pod| pod.metadata.name.clone().unwrap())
+                            .collect::<Vec<_>>()
+                    );
+                    trace!(
+                        "labels: [{:?}]",
+                        get_node_and_group_labels(role_group, &node_type)
+                    );
+                    let nodes_that_need_pods = k8s_utils::find_nodes_that_need_pods(
+                        nodes,
+                        &self.existing_pods,
+                        &get_node_and_group_labels(role_group, &node_type),
+                    );
+
+                    for node in nodes_that_need_pods {
+                        let node_name = if let Some(node_name) = &node.metadata.name {
```

Comment:
Just a personal preference to keep the error handling up top:
```
if node.metadata.name.is_none() {
    warn!("No name found in metadata, this should not happen! Skipping node: [{:?}]", node);
    continue;
}

let node_name = node.metadata.name.as_deref().unwrap();
```

Reply:
I think I prefer the `if let `, as this would otherwise be a call to .unwrap() which would keep showing up in searches for "leftover" unwraps even though it is perfectly safe.